File: src/core/measurement_exporter.py
"""Export measurement data to Excel."""

from __future__ import annotations

from pathlib import Path

import pandas as pd

from src.models.project import Project


def export_measurements_excel(project: Project, output_path: str) -> None:
    """Write all measurements across all stations to a single Excel file."""
    rows = []
    for station in project.stations:
        for ann in station.annotations:
            image_name = Path(ann.image_path).name
            calibrated = ann.scale_factor > 1.0
            calib_info = (
                f"{ann.scale_factor:.2f} px/{ann.scale_unit}"
                if calibrated else "uncalibrated"
            )
            unit = ann.scale_unit if calibrated else "px"
            unit2 = f"{unit}²"

            for m in ann.measurements:
                is_area = m.type == "polygon"
                rows.append({
                    "Station": station.name,
                    "Image": image_name,
                    "Calibration": calib_info,
                    "Fragment Label": m.label,
                    "Spesies/Genus": m.species,
                    "Measurement Type": m.type,
                    f"Width ({unit})": round(m.auto_width, 4) if m.auto_width else "",
                    f"Height ({unit})": round(m.auto_height, 4) if m.auto_height else "",
                    f"Area ({unit2})": round(m.area, 4) if is_area else "",
                    f"Perimeter ({unit})": round(m.perimeter_len, 4) if is_area else "",
                    f"Value": round(m.value, 4),
                    "Unit": unit2 if is_area else unit,
                    "ID": m.id,
                })

    columns = [
        "Station", "Image", "Calibration",
        "Fragment Label", "Spesies/Genus", "Measurement Type",
        f"Width ({unit})", f"Height ({unit})",
        f"Area ({unit2})", f"Perimeter ({unit})",
        "Value", "Unit", "ID",
    ]
    # fall back to generic headers when no rows exist
    if not rows:
        columns = [
            "Station", "Image", "Calibration",
            "Fragment Label", "Spesies/Genus", "Measurement Type",
            "Width", "Height", "Area", "Perimeter",
            "Value", "Unit", "ID",
        ]

    df = pd.DataFrame(rows) if rows else pd.DataFrame(columns=columns)

    with pd.ExcelWriter(output_path, engine="openpyxl") as writer:
        df.to_excel(writer, sheet_name="Measurements", index=False)

        ws = writer.sheets["Measurements"]
        for col in ws.columns:
            max_len = max((len(str(cell.value or "")) for cell in col), default=8)
            ws.column_dimensions[col[0].column_letter].width = min(max_len + 4, 50)

File: rust/src/core/measurement_exporter.rs
//! Export measurement data to Excel.

use std::path::Path;

use anyhow::Result;
use rust_xlsxwriter::{Workbook, Worksheet};

use crate::core::table::{column_union, Cell, Row};
use crate::models::Project;

/// Write all measurements across all stations to a single Excel file.
pub fn export_measurements_excel(project: &Project, output_path: impl AsRef<Path>) -> Result<()> {
    let mut rows: Vec<Row> = Vec::new();

    for station in &project.stations {
        for ann in &station.annotations {
            let image_name = Path::new(&ann.image_path)
                .file_name()
                .map(|n| n.to_string_lossy().into_owned())
                .unwrap_or_else(|| ann.image_path.clone());
            let calibrated = ann.scale_factor > 1.0;
            let calib_info = if calibrated {
                format!("{:.2} px/{}", ann.scale_factor, ann.scale_unit)
            } else {
                "uncalibrated".to_string()
            };
            let unit = if calibrated { ann.scale_unit.clone() } else { "px".to_string() };
            let unit2 = format!("{unit}\u{b2}");

            for m in &ann.measurements {
                let is_area = m.kind == "polygon";
                let row: Row = vec![
                    ("Station".into(), station.name.clone().into()),
                    ("Image".into(), image_name.clone().into()),
                    ("Calibration".into(), calib_info.clone().into()),
                    ("Fragment Label".into(), m.label.clone().into()),
                    ("Spesies/Genus".into(), m.species.clone().into()),
                    ("Measurement Type".into(), m.kind.clone().into()),
                    (
                        format!("Width ({unit})"),
                        if m.auto_width != 0.0 { round4(m.auto_width).into() } else { Cell::None },
                    ),
                    (
                        format!("Height ({unit})"),
                        if m.auto_height != 0.0 { round4(m.auto_height).into() } else { Cell::None },
                    ),
                    (
                        format!("Area ({unit2})"),
                        if is_area { round4(m.area).into() } else { Cell::None },
                    ),
                    (
                        format!("Perimeter ({unit})"),
                        if is_area { round4(m.perimeter_len).into() } else { Cell::None },
                    ),
                    ("Value".into(), round4(m.value).into()),
                    ("Unit".into(), if is_area { unit2.clone() } else { unit.clone() }.into()),
                    ("ID".into(), m.id.clone().into()),
                ];
                rows.push(row);
            }
        }
    }

    // Fall back to generic headers when no measurements exist at all.
    let columns = if rows.is_empty() {
        vec![
            "Station", "Image", "Calibration", "Fragment Label", "Spesies/Genus",
            "Measurement Type", "Width", "Height", "Area", "Perimeter", "Value", "Unit", "ID",
        ]
        .into_iter()
        .map(String::from)
        .collect::<Vec<_>>()
    } else {
        column_union(&rows)
    };

    let mut workbook = Workbook::new();
    let sheet: &mut Worksheet = workbook.add_worksheet();
    sheet.set_name("Measurements")?;

    for (col, name) in columns.iter().enumerate() {
        sheet.write_string(0, col as u16, name.as_str())?;
    }
    for (r, row) in rows.iter().enumerate() {
        let lookup: std::collections::HashMap<&str, &Cell> =
            row.iter().map(|(k, v)| (k.as_str(), v)).collect();
        for (col, name) in columns.iter().enumerate() {
            if let Some(cell) = lookup.get(name.as_str()) {
                write_cell(sheet, r as u32 + 1, col as u16, cell)?;
            }
        }
    }

    sheet.autofit();
    workbook.save(output_path)?;
    Ok(())
}

fn write_cell(sheet: &mut Worksheet, row: u32, col: u16, cell: &Cell) -> Result<()> {
    match cell {
        Cell::Str(s) => {
            sheet.write_string(row, col, s.as_str())?;
        }
        Cell::Num(v) => {
            sheet.write_number(row, col, *v)?;
        }
        Cell::Int(v) => {
            sheet.write_number(row, col, *v as f64)?;
        }
        Cell::None => {}
    }
    Ok(())
}

fn round4(v: f64) -> f64 {
    (v * 10000.0).round() / 10000.0
}
